Mathlib/Algebra/Order/Floor.lean

```diff
@@ -0,0 +1,1689 @@
+#align round_int_add round_int_add
+
+@[simp]
+theorem round_nat_add (x : α) (y : ℕ) : round ((y : α) + x) = y + round x := by
+  rw [add_comm, round_add_nat, add_comm]
+#align round_nat_add round_nat_add
+
+theorem abs_sub_round_eq_min (x : α) : |x - round x| = min (fract x) (1 - fract x) := by
+  simp_rw [round, min_def_lt, two_mul, ← lt_tsub_iff_left]
+  cases' lt_or_ge (fract x) (1 - fract x) with hx hx
+  · rw [if_pos hx, if_pos hx, self_sub_floor, abs_fract]
+  · have : 0 < fract x := by
+      replace hx : 0 < fract x + fract x := lt_of_lt_of_le zero_lt_one (tsub_le_iff_left.mp hx)
+      simpa only [← two_mul, zero_lt_mul_left, zero_lt_two] using hx
+    rw [if_neg (not_lt.mpr hx), if_neg (not_lt.mpr hx), abs_sub_comm, ceil_sub_self_eq this.ne.symm,
+      abs_one_sub_fract]
+#align abs_sub_round_eq_min abs_sub_round_eq_min
+
+theorem round_le (x : α) (z : ℤ) : |x - round x| ≤ |x - z| := by
+  rw [abs_sub_round_eq_min, min_le_iff]
+  rcases le_or_lt (z : α) x with (hx | hx) <;> [left, right]
+  · conv_rhs => rw [abs_eq_self.mpr (sub_nonneg.mpr hx), ← fract_add_floor x, add_sub_assoc]
+    simpa only [le_add_iff_nonneg_right, sub_nonneg, cast_le] using le_floor.mpr hx
+  · rw [abs_eq_neg_self.mpr (sub_neg.mpr hx).le]
+    conv_rhs => rw [← fract_add_floor x]
+    rw [add_sub_assoc, add_comm, neg_add, neg_sub, le_add_neg_iff_add_le, sub_add_cancel,
+      le_sub_comm]
+    norm_cast
+    exact floor_le_sub_one_iff.mpr hx
+#align round_le round_le
+
+end LinearOrderedRing
+
+section LinearOrderedField
+
+variable [LinearOrderedField α] [FloorRing α]
+
+theorem round_eq (x : α) : round x = ⌊x + 1 / 2⌋ := by
+  simp_rw [round, (by simp only [lt_div_iff', two_pos] : 2 * fract x < 1 ↔ fract x < 1 / 2)]
+  cases' lt_or_le (fract x) (1 / 2) with hx hx
+  · conv_rhs => rw [← fract_add_floor x, add_assoc, add_left_comm, floor_int_add]
+    rw [if_pos hx, self_eq_add_right, floor_eq_iff, cast_zero, zero_add]
+    constructor
+    -- Porting note: was linarith [fract_nonneg x]
+    · refine add_nonneg (fract_nonneg _) ?_
+      refine div_nonneg one_pos.le (le_of_lt ?_)
+      exact two_pos
+    · rw [←lt_sub_iff_add_lt]
+      convert hx
+      rw [sub_eq_iff_eq_add, add_halves]
+  · have : ⌊fract x + 1 / 2⌋ = 1 := by
+      rw [floor_eq_iff]
+      constructor
+      -- was norm_num -- <;> linarith [fract_lt_one x]
+      · rw [←sub_le_iff_le_add]
+        convert hx
+        rw [sub_eq_iff_eq_add, add_halves, cast_one]
+      · refine (add_lt_add_right (fract_lt_one x) (1 / 2)).trans ?_
+        rw [cast_one]
+        apply add_lt_add_left
+        conv_rhs => rw [←add_halves 1]
+        norm_num
+    rw [if_neg (not_lt.mpr hx), ← fract_add_floor x, add_assoc, add_left_comm, floor_int_add,
+      ceil_add_int, add_comm _ ⌊x⌋, add_right_inj, ceil_eq_iff, this, cast_one, sub_self]
+    constructor
+    -- was linarith [fract_lt_one x]
+    · apply lt_of_lt_of_le _ hx
+      norm_num
+    · exact (fract_lt_one x).le
```

Comment:
Replacing those `linarith` calls makes me sad

Reply:
Opened an issue for one of them:
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/linarith.20failure

Reply:
Now that that issue has been fixed (#1358), I believe all remaining linarith failures are due to the lack of fraction functionality in `ring` (#1189).  I rewrote your fixes on the broken linarith calls in order to make this transparent.